src/ngraph/runtime/gpu/gpu_compiled_function.cpp

```diff
@@ -0,0 +1,161 @@
+#include "ngraph/runtime/gpu/pass/gpu_batch_norm_cache.hpp"
+#include "ngraph/runtime/gpu/pass/gpu_layout.hpp"
+#include "ngraph/runtime/gpu/pass/gpu_rnn_fusion.hpp"
+#include "ngraph/runtime/gpu/pass/tensor_memory_reservation.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+const std::string runtime::gpu::GPU_CompiledFunction::s_output_dir = "gpu_codegen";
+static std::mutex s_compilation;
+
+class GPUStaticInitializers
+{
+public:
+    GPUStaticInitializers()
+    {
+        file_util::remove_directory(runtime::gpu::GPU_CompiledFunction::s_output_dir);
+        file_util::make_directory(runtime::gpu::GPU_CompiledFunction::s_output_dir);
+    }
+};
+
+static GPUStaticInitializers s_static_initializers;
+
+const size_t runtime::gpu::GPU_CompiledFunction::GPU_CompiledFunction::s_memory_pool_alignment = 64;
+
+runtime::gpu::GPU_CompiledFunction::GPU_CompiledFunction(
+    const shared_ptr<ngraph::Function>& function,
+    std::shared_ptr<GPU_Backend::BackendContext>& shared_context)
+    : m_runtime(nullptr)
+    , m_function(function)
+    , m_emit_timing(false)
+    , m_is_compiled(false)
+    , m_shared_context(shared_context)
+{
+}
+
+runtime::gpu::GPU_CompiledFunction::~GPU_CompiledFunction()
+{
+}
+
+std::shared_ptr<runtime::gpu::GPU_CompiledFunction> runtime::gpu::GPU_CompiledFunction::make(
+    const std::shared_ptr<ngraph::Function>& function,
+    std::shared_ptr<GPU_Backend::BackendContext>& shared_context)
+{
+#if defined(NGRAPH_DEX_ONLY)
+    return std::make_shared<runtime::gpu::GPU_InternalFunction>(function, shared_context);
+#else
+    // For now codegen is default unless explicitly disabled
+    bool use_codegen = true;
+    if (auto env = std::getenv("NGRAPH_CODEGEN"))
+    {
+        std::string env_codegen(env);
+        if (env_codegen == "0" || env_codegen == "false" || env_codegen == "False" ||
+            env_codegen == "FALSE" || env_codegen == "no" || env_codegen == "No" ||
+            env_codegen == "NO")
+        {
+            use_codegen = false;
+        }
+    }
+    if (use_codegen)
+    {
+        return std::make_shared<runtime::gpu::GPU_ExternalFunction>(function, shared_context);
+    }
+    else
+    {
+        return std::make_shared<runtime::gpu::GPU_InternalFunction>(function, shared_context);
+    }
+#endif
+}
+
+void runtime::gpu::GPU_CompiledFunction::compile()
+{
+    if (m_is_compiled)
+    {
+        return;
+    }
+    std::unique_lock<std::mutex> lock(s_compilation);
+
+    m_function_name = m_function->get_name();
+
+    auto allocator = std::make_shared<runtime::gpu::GPUAllocator>(
+        m_shared_context->m_primitive_emitter->get_memory_allocator());
+
+    ngraph::pass::Manager pass_manager;
+#if CUDNN_VERSION >= 7200
+    // recurrent network fusion
+    pass_manager.register_pass<runtime::gpu::pass::LSTMFusion>();
+    pass_manager.register_pass<runtime::gpu::pass::RNNFusion>();
+    pass_manager.register_pass<ngraph::pass::AlgebraicSimplification>();
```

Comment:
nit: 
`pass_manager.register_pass<ngraph::pass::AlgebraicSimplification>();` can be move out of the `#if`.

Reply:
or it should stay as is if there is order requirement.

Reply:
There is an order requirement for the RNN fusions currently.